runtime/src/recovery_flow.rs

```diff
@@ -0,0 +1,67 @@
+/*++
+
+Licensed under the Apache-2.0 license.
+
+File Name:
+
+    recovery_flow.rs
+
+Abstract:
+
+    File contains the implementation of the recovery flow for MCU firmware and SoC manifest.
+
+--*/
+
+use crate::Drivers;
+use caliptra_auth_man_types::{
+    AuthManifestImageMetadataCollection, AuthManifestPreamble, AuthorizationManifest,
+    AUTH_MANIFEST_PREAMBLE_SIZE,
+};
+use caliptra_cfi_derive_git::{cfi_impl_fn, cfi_mod_fn};
+use caliptra_drivers::{AxiAddr, Dma, DmaReadTarget, DmaReadTransaction, DmaRecovery};
+use caliptra_kat::{CaliptraError, CaliptraResult};
+use caliptra_registers::i3ccsr::RegisterBlock;
+use core::{
+    any::TypeId,
+    cell::{Cell, RefCell},
+};
+use ureg::{Mmio, MmioMut, Uint, UintType};
+use zerocopy::{AsBytes, FromBytes};
+
+pub enum RecoveryFlow {}
+
+impl RecoveryFlow {
+    /// Load the SoC Manifest and MCU firwmare
+    #[cfg_attr(not(feature = "no-cfi"), cfi_impl_fn)]
+    pub(crate) fn recovery_flow(drivers: &mut Drivers) -> CaliptraResult<()> {
+        const SOC_MANIFEST_INDEX: u32 = 1;
+        const MCU_FIRMWARE_INDEX: u32 = 2;
+
+        let dma = &drivers.dma;
+        let dma_recovery =
+            DmaRecovery::new(drivers.soc_ifc.recovery_interface_base_addr().into(), dma);
+
+        // // download SoC manifest
+        let _soc_size_bytes = dma_recovery.download_image_to_mbox(SOC_MANIFEST_INDEX)?;
+        let Some(preamble) = AuthManifestPreamble::read_from_prefix(drivers.mbox.raw_mailbox_contents()) else {
+            return Err(CaliptraError::IMAGE_VERIFIER_ERR_MANIFEST_SIZE_MISMATCH);
+        };
+        let Some(image_metadata_collection) = AuthManifestImageMetadataCollection::read_from_prefix(&drivers.mbox.raw_mailbox_contents()[AUTH_MANIFEST_PREAMBLE_SIZE..]) else {
+            return Err(CaliptraError::IMAGE_VERIFIER_ERR_MANIFEST_SIZE_MISMATCH);
+        };
```

Comment:
nit: can probably simplify this by making a single stuct that contains `AuthManifestPreamble` and `AuthManifestImageMetadataCollection` and then doing a single read.